### DISCUSSION TITLE
Could we reduce boilerplate with sensible defaults?

### DISCUSSION THREAD

**Discussion by aaaidan** [HomeSpan/HomeSpan#32] [Ideas]
What if homeSpan provided a "best effort" default implementation of `AccessoryInformation` and `HAPProtocolInformation`? Perhaps if the user provides their own, the built-in implementation is discarded.

HomeSpan already hugely reduces the amount of code and effort required to get a HomeKit accessory running on a network. There are lots of reasons for this, but it's no small part due to its expressive API, and the way it abstracts HAP. 

However, there are still "bureaucratic" parts of HAP that leak through this abstraction: e.g. the `AccessoryInformation` and `HAPProtocolInformation`. This is most obvious on very simple or elementary cases, like a light switch or a contact sensor, where it feels like there is as much boilerplate as there is application code. Despite homeSpan's great docs and tutorials, this boilerplate is also, I think, a small but unnecessary barrier to entry. 

The `homeSpan.begin` function already gathers some of the required chars for `AccessoryInformation` in optional parameters. Could mandatory protocol information be gathered via mandatory params? 

Here's one way it could look for a very minimal simple contact sensor:

```
#include <Arduino.h>
#include <HomeSpan.h>

SpanCharacteristic* contactSensorState;
const unit8_t SENSOR_PIN = 23;

void setup() {
  Serial.begin(115200);
  pinMode(PIN_CONTACT_SENSOR, INPUT_PULLUP);

  homeSpan.begin(
    "Contact Sensor" // generic name for the device, used for displayName, model, etc (mandatory)
    Category::Sensors,  // category (mandatory?)
  );

  new Service::ContactSensor();
  sensorState = new Characteristic::ContactSensorState();
}

void loop() {
  homeSpan.poll();
  auto val = digitalRead(SENSOR_PIN) == 0 ? ContactSensorState.CLOSED : ContactSensorState.OPEN; // values for HAP 1.1.0
  sensorState->setVal(val);
}
```

Notice there's no `new SpanAccessory()` – a default one is provided by homeSpan if it's not specified.

In this case, the accessory database could look like:

```
Accessory
  AccessoryInformation
    Name: Contact Sensor
    Manufacturer: HomeSpan
    SerialNumber: 92817346 [mac address, or ESP32 serial number]
    Model: Contact Sensor
    FirmwareRevision: 0.1
    Identify: [default implementation blinks "status" LED]
  HAPProtocolInformation
    Version: 1.1.0
  ContactSensor
    ContactSensorState
```

(Maybe there are other places homeSpan could take a good best-guess. Could the category be safely inferred if there's only one `Service`? At the moment, it defaults to lighting regardless of what services are published, so a good faith effort seems like it wouldn't make things worse (protocol wise), and would reduce boilerplate and ease the on-ramp.)

Maybe it's important for HomeSpan to directly mirror the homekit accessory database, and that trying to "paper over" HAP's bureaucratic chattiness isn't in line with HomeSpan's goals. Concerns about "magic" library behaviour, bloat, and exceptions are also reasonable. But looking at examples and tutorials, ease of use and expressiveness seem to be a goals of HomeSpan, and I think taming HAP could make HomeSpan even more of a joy to use.

Changes like this would probably require in a major semvar bump, but maybe with some care it could be backward compatible.

Thoughts?

**Comment by HomeSpan** [comment]
These are some very good ideas, though you correctly surmised that:

> ... it's important for HomeSpan to directly mirror the homekit accessory database, and that trying to "paper over" HAP's bureaucratic chattiness isn't in line with HomeSpan's goals. Concerns about "magic" library behaviour, bloat, and exceptions are also reasonable.

I think what makes HomeSpan robust is that it does indeed directly mirror HAP, and for complicated configurations that don't exactly follow HAP you have the ability to see and control exactly what HomeKit is doing (so you can work around HomeKit as needed with some additional logic).

However, I think a potential solution to might be to create a "macro library" that can form an overlay of the HomeSpan methods.  For example, a simple "SETUP" macro can take a few arguments, generate the homeSpan.begin(), the initial Accessory Info, HAPProtocolInfo, etc., and leave the user ready for specifying Services and Characteristics.  The user would then have the option of creating a new Accessory, but only if a second one was needed.

What's nice about the macro method is that is forms a clean layer between the base code and easy-to-create "convenience" methods that can be combined and re-combined in as many ways as needed.

**Reply by aaaidan** [reply]
Helpful to know that your vision for homeSpan is to mostly respect and align with HAP, even where it's a bit unergonomic. I think this is a really solid principle.

And yes, I like this idea of a "convenience" layer. I think there is room for it. It would be great if there was a smooth onramp from it to the existing "full" API, when you outgrow it.

Something that a basic macro implementation probably can't tackle is auto-selecting the category in the common case that there is only one user-defined service type. Maybe this seems like an unwise feature, and I don't think this is a make-or-break problem to solve but, to me, it always has seemed a strange/sloppy part of HAP. I may just be missing why it is necessary.

Also, one thing homeSpan might want to implement sometime is constants for characteristic states. Like `ContactSensor::CLOSED` etc.  I think this would be a necessary part of injecting the HAP version "in good faith". Do you have a plan for this? Sounds like it could be a boring low-skill task in places. Could I help do any heavy lifting?

**Comment by HomeSpan** [comment]
Regarding the Category, note that it is purely cosmetic and not use by HAP.  It's only used to display a (potentially) helpful icon during the pairing process.  After the pairing process is complete, the Home App ignores it.  The actual tile icons in Home App are set according to the Services selected (specifically, the primary operating service if there is more than one service in an Accessory).  You can create a Fan Service but still use LightBulb for the category - the category had no effect.  The Services don't necessarily map one-to-one with Categories, which is why I think HAP keeps them as separate concepts, giving users the freedom to choose any combination, or just set the category to "Other" no matter what Services are used.

When I first started creating HomeSpan I was planning on having macros for each characteristic as you suggest.  However, the problem I ran into was that it was not obvious what the names of the constants should be.  I would therefore have to create document listing each constant AND explaining what they mean and how they are used.  But that document already exists - it's the HAP R2 guide.

My expectation was that every user would download that from Apple, since it contains a list of all the Services and Characteristics, and most importantly, what they mean, how they are used, and what the exceptions are.  HAP does not seem to use a common set of descriptions for the Characteristics, so I figured it would only create more confusion if I overlaid a new nomenclature.  Once you have the HAP R2 doc, you can see whether val=0 means open, or val=1 is for open, etc.

That said, there is definitely more to do to improve robustness, especially around error-checking.  HomeSpan provides no error-checking when you update a value with setVal().  It probably should prevent incorrect values from being sent (and raise a warning when that happens) since a bad value will cause a NO RESPONSE error on Controllers that try to connect to HomeSpan.

I'll look into adding a generic method to specify the valid ranges for each characteristic from within the code (likely an extension to HapCharList in HAPConstants.h, with appropriate additions to the SpanCharacteristic structure, and of course new logic to range-check inside setVal).  I should be able to tackle this over the weekend.  Once done, if you're interested and feel like helping populate the ranges based on the HAP R2 doc, that would be great.
